[graphql/telemetry] Add metrics and tracing

stefan-mysten opened the pull request:
## Description 

This PR adds a number of metrics and logging. See last comment for more detailed information (after several updates).

## Test plan

Mostly manual, firing up a service and a Grafana dashboard to test out the prometheus metrics.

### Metrics
<img width="1720" alt="image" src="https://github.com/MystenLabs/sui/assets/135084671/b825f365-ce41-4d2e-a77f-a46ebac561b8">


### Logging
**Query**
```graphql
{
  transactionBlocks(first:400) {
    nodes {
      digest
    }
  }
}
```
**Log**
<img width="1720" alt="image" src="https://github.com/MystenLabs/sui/assets/135084671/f7d394be-1883-4d8f-9e78-929462f6d409">

**Query**
```graphql
{
  epoch(id: -1) {
    epochId
  }
}
```
**Log** - note that for those errors that we do not add an `error_code` we are returning a `warn` log.
<img width="1720" alt="image" src="https://github.com/MystenLabs/sui/assets/135084671/0e0d129c-9b99-4be0-83ca-9635c8cfd475">




---
If your changes are not user-facing and do not break anything, you can skip the following section. Otherwise, please briefly describe what has changed under the Release Notes section.

### Type of Change (Check all that apply)

- [ ] protocol change
- [ ] user-visible impact
- [ ] breaking change for a client SDKs
- [ ] breaking change for FNs (FN binary must upgrade)
- [ ] breaking change for validators or node operators (must upgrade binaries)
- [ ] breaking change for on-chain data layout
- [ ] necessitate either a data wipe or data migration

### Release notes

stefan-mysten commented on crates/sui-graphql-rpc/src/extensions/query_limits_checker.rs:
```diff
@@ -135,11 +136,28 @@ impl Extension for QueryLimitsChecker {
+            x.request_metrics
+                .query_validation_latency_by_path
+                // TODO make the path here
+                .with_label_values(&["path"])
```
This path is easy to get for errors, but could not figure out yet how to do it for normal queries. Also, it's not clear to me how to make a path of a very complex query (multiple nested nodes, multiple sub-level queries).

wlmyng replied:
Errors in what sense?

We can extend the queue to also take in a string of its parents like "objects[n].owner.coins[n]"

stefan-mysten replied:
Leaving it out for now.

stefan-mysten commented on crates/sui-graphql-rpc/src/extensions/query_limits_checker.rs:
```diff
@@ -135,11 +136,28 @@ impl Extension for QueryLimitsChecker {
+                metrics
+                    .request_metrics
+                    .query_payload_error
+                    .with_label_values(&["validation"])
```
the label does not look right, we probably don't want a label here.

wlmyng replied:
Me noob: what do labels do?

stefan-mysten replied:
Labels are a way to group values by some key. E.g., you could use a status label, and then have the values 200, 400, 500, each with its related values.

wlmyng commented on crates/sui-graphql-rpc/src/data/pg.rs:
```diff
@@ -37,13 +44,20 @@ impl QueryExecutor for PgManager_ {
     {
         let max_cost = self.limits.max_db_query_cost;
-        self.inner
+        let instant = Instant::now();
```
When @amnn's refactor lands, would we want to track the time of the whole transaction, but should be a simple lift over to the execute and execute_repeatable methods

stefan-mysten replied:
Yes, there will be some changes once that lands. I am still playing with it but I think I got the gist.

wlmyng commented on crates/sui-graphql-rpc/src/extensions/query_limits_checker.rs:
```diff
@@ -185,9 +203,16 @@ impl Extension for QueryLimitsChecker {
+        if let Some(timer) = latency_timer {
+            timer.observe_duration();
+        }
+        if let Some(timer) = timer_latency_by_path {
+            timer.observe_duration();
+        }
```
What's the difference between this and the other latency timer? They get instantiated and updated at the same level - I was assuming that we update the path timer within the for loop or something

stefan-mysten replied:
The high level idea was to record a metric by query path (some sort of normalized path), similarly to what we do today with JSON RPC (each endpoint has its own "metric" values).

wlmyng commented on crates/sui-graphql-rpc/src/extensions/query_limits_checker.rs:
```diff
@@ -326,8 +360,16 @@ impl QueryLimitsChecker {
+    limits: &Limits,
+    cost: &ComponentCost,
+    pos: Option<Pos>,
+    ctx: &ExtensionContext<'_>,
```
Can we pass in just the metrics instead of the entire extension context

stefan-mysten replied:
That makes sense!

stefan-mysten replied:
We need this to get the `query id` and the `session_id`

wlmyng commented on crates/sui-graphql-rpc/src/extensions/query_limits_checker.rs:
```diff
@@ -348,6 +394,11 @@ fn check_limits(limits: &Limits, cost: &ComponentCost, pos: Option<Pos>) -> Serv
 
     if cost.output_nodes > limits.max_output_nodes {
+        if let Some(metrics) = ctx.data_opt::<Metrics>() {
+            // TODO is this a client error though?
```
Imo yes because a caller can calculate the number of output nodes their query will produce. If it's not a client error then the onus is on us to fix

stefan-mysten replied:
Great, thank you!

wlmyng commented on crates/sui-graphql-rpc/src/metrics.rs:
```diff
@@ -1,15 +1,209 @@
+    /// Updates the DB related metrics (latency, error, success)
+    pub(crate) fn observe_db_data(&self, db_latency: f64, succeeded: bool) {
+        self.db_metrics
+            .db_fetch_latency
+            .with_label_values(&["latency", "db"])
```
This seems redundant, are there other labels we plan to have under the db_fetch_latency metric?

stefan-mysten replied:
This is wrong and I changed it now.

wlmyng commented on crates/sui-graphql-rpc/src/metrics.rs:
```diff
@@ -1,15 +1,209 @@
+    }
+
+    /// Use this function to increment the number of errors due to bad input
+    pub(crate) fn inc_num_bad_input_errors(&self, label: &str) {
```
Could probs be named inc_input_errors

stefan-mysten replied:
This also seems to be wrong the way I did it, so now it will be refactored. Thanks though!

stefan-mysten replied:
Renamed to `inc_errors` which works with all the errors and groups them by `path`.

wlmyng commented on crates/sui-graphql-rpc/src/metrics.rs:
```diff
@@ -1,15 +1,209 @@
+                    if let async_graphql_value::Value::String(val) = code.clone().into_value() {
+                        if val == code::INTERNAL_SERVER_ERROR {
+                            self.request_metrics
+                                .num_internal_errors_by_path
```
Why not have the one metric, like request_error, and then multiple labels, bad_input_error, internal_error etc

stefan-mysten replied:
Ah yes, that's where I ended up in the end.

wlmyng commented on crates/sui-graphql-rpc/src/metrics.rs:
```diff
@@ -1,15 +1,209 @@
+        self.request_metrics
+            .num_bad_input_errors_by_path
+            // TODO when measuring cost related errors, there's no path
+            // not 100% how to label such an error?
```
Imo we don't rly need all the details on a bad input error since the fault is on the client and they can resolve on their end. So it's fine if we don't have path info

wlmyng commented on crates/sui-graphql-rpc/src/metrics.rs:
```diff
@@ -1,15 +1,209 @@
+    /// The path is the same as the path that the user sees.
+    pub(crate) fn inc_errors(&self, errors: Vec<ServerError>) {
+        for err in errors {
+            if let Some(ext) = err.extensions {
+                if let Some(code) = ext.get("code") {
+                    if let async_graphql_value::Value::String(val) = code.clone().into_value() {
+                        if val == code::INTERNAL_SERVER_ERROR {
```
What are your thoughts on splitting out these nested ifs into

let Some(ext) = err.extensions else continue

let Some(code) = ...

Yes we have a lot of continue but it's easier on the eyes since we don't have such deep nesting. You can also chain Some, like let Some(Some... 

I'm guessing code is a GqlValue and you're converting it into a ConstValue

But then a bit confused about the val == code::INTERNAL_SERVER_ERROR, how does that work?

stefan-mysten replied:
`ext.get("code")` returns a `&ConstValue`, so need to call `into_value()` on it to be able to match on `Value::String`. Otherwise it complains that it cannot do a comparison between `Value` and `String`.

wlmyng replied:
i dont follow why we are comparing val against itself:

if let async_graphql_value::Value::String(val) =
                                code.clone().into_value()
                            {
                                if val == code::INTERNAL_SERVER_ERROR {

unless I'm not understanding it correctly? val is from code.clone().into_value(). then we compare val against code::INTERNAL_SERVER_ERROR? 

oh. is code from error::code? If so, can we avoid using the same word?

stefan-mysten replied:
Sorry for confusing you. The high level goal is to match on the error code that we're setting in the `error.rs` file. I will change `code` to `error_code`, and import the error codes constants so that we have more like this or better maybe I can do some maps to get the set error code as a string:
```
            if let Some(ext) = err.extensions {
                if let Some(error_code) = ext.get("code") {
                    if let async_graphql_value::Value::String(val) = error_code.clone().into_value() {
                        if val == INTERNAL_SERVER_ERROR {
```

wlmyng replied:
> ext.get("code") returns a &ConstValue, so need to call into_value() on it to be able to match on Value::String. Otherwise it complains that it cannot do a comparison between Value and String.

Oddly enough, when I try the following code snippet, I do not get the error. Also, I'd think that one would be able to match a reference to a ConstValue to async_graphql::Value::String just like a reference to any other enum. 

    if let Some(code) = ext.get("code") {
        if let async_graphql::Value::String(val) = code {
            if val == code::BAD_USER_INPUT {
                println!("oops");
            }
        }
    }

stefan-mysten replied:
Admittedly, I did not give it a lot of thought. I'll look at it again and try to simplify.

wlmyng commented on crates/sui-graphql-rpc/src/metrics.rs:
```diff
@@ -1,15 +1,209 @@
+        }
+    }
+
+    /// Use this function to increment the number of request error.
+    /// If you need to increase error rate due to bad input, use both
+    /// this function and `inc_error_bad_input`
+    pub(crate) fn inc_num_query_success(&self) {
```
The name of the function is query_success though. I also feel like an aggregate request error count isn't that useful. I can just sum the count of bad input and internal errors over a window on grafana

stefan-mysten replied:
> I can just sum the count of bad input and internal errors over a window on grafana
I'll redo this, thanks!

wlmyng commented on crates/sui-graphql-rpc/src/metrics.rs:
```diff
@@ -1,15 +1,209 @@
+impl DBMetrics {
+    pub(crate) fn new(registry: &Registry) -> Self {
+        Self {
+            db_num_fetches: register_int_counter_vec_with_registry!(
+                "db_fetch",
```
I feel like these should be as similar as possible (or the same)

wlmyng commented on crates/sui-graphql-rpc/src/metrics.rs:
```diff
@@ -1,15 +1,209 @@
+        Self {
+            db_num_fetches: register_int_counter_vec_with_registry!(
+                "db_fetch",
+                "The total number of DB requests grouped by success or error",
```
Imo success count is not useful, you usually end up ignoring it

stefan-mysten replied:
I am still learning the best practices here, so happy to hear other folks' experience. We can either do `db_fetch` as number of total DB requests, and then `db_fetch_error` to understand the error rate, or we can have them as `success` and `error` which is the total num of db fetches, from which we can also compute the error. I thought one field instead of two might be easier.

wlmyng commented on crates/sui-graphql-rpc/src/metrics.rs:
```diff
@@ -62,10 +261,68 @@ impl RequestMetrics {
+                registry,
+            )
+            .unwrap(),
+            query_latency: register_histogram_with_registry!(
```
Is this inclusive of the validation latency?

stefan-mysten replied:
Yes

stefan-mysten replied:
I actually lost my mind on these, so we do have query validation latency.

amnn commented on crates/sui-graphql-rpc/src/data/pg.rs:
```diff
@@ -124,9 +152,19 @@ mod query_cost {
+                cost,
+                max_db_query_cost,
+                exceeds = true,
+                "[Cost] Estimated cost"
```
There's no need to make this change.  The reason this code is wrapped in a `query_cost` module is to make sure it gets annotated with "query_cost" in its target.

stefan-mysten replied:
True, but most of the others have a little string preview "[Cost]" so I added it to keep it consistent. I'll review it once it's close to merging.

stefan-mysten replied:
Removed.

amnn commented on crates/sui-graphql-rpc/src/extensions/query_limits_checker.rs:
```diff
@@ -135,11 +137,15 @@ impl Extension for QueryLimitsChecker {
             .expect("No service config provided in schema data");
 
         if query.len() > cfg.limits.max_query_payload_size as usize {
+            if let Some(metrics) = ctx.data_opt::<Arc<Metrics>>() {
```
Should we also observe the query length in this case?

stefan-mysten replied:
Why not?

stefan-mysten replied:
Added!

amnn commented on crates/sui-graphql-rpc/src/metrics.rs:
```diff
@@ -1,72 +1,222 @@
+        self.request_metrics.query_latency.observe(time);
+    }
+
+    /// Record the time needed for validation the query
```
```suggestion
    /// Record the time needed for validating the query
```

amnn commented on crates/sui-graphql-rpc/src/metrics.rs:
```diff
@@ -1,72 +1,222 @@
-    1., 2., 4., 8., 12., 16., 24., 32., 48., 64., 96., 128., 256., 512., 1024.,
-];
+#[derive(Clone)]
+pub(crate) struct RequestMetrics {
```
We should also track the number of requests in flight -- this is usually done by counting the number of queries as they come into the system and as they leave the system.

stefan-mysten replied:
Sounds good!

amnn commented on crates/sui-graphql-rpc/src/server/builder.rs:
```diff
@@ -636,40 +658,96 @@ pub mod tests {
-        assert_eq!(metrics2.input_nodes.get_sample_sum(), 1.);
-        assert_eq!(metrics2.output_nodes.get_sample_sum(), 1.);
-        assert_eq!(metrics2.query_depth.get_sample_sum(), 1.);
+        /// changing this to Histogram (mysten-metrics) rather than prometheus::Histogram makes it very difficult to test.
```
Why is that?

stefan-mysten replied:
`prometheus` offered a simple `get_sample_count()` method to get the number of items in the histogram. Our histogram does not have that, and it's a bit more involved due to the automated buckets (https://github.com/MystenLabs/sui/blob/main/crates/mysten-metrics/src/histogram.rs#L371-L419)

I'll probably try to submit a PR to have a few convenience methods that would help with testing, just did not get around it yet.

stefan-mysten replied:
I am unable to get this working properly, so disabled the tests for now (manually checked that they work) and asking folks about it.

amnn reviewed (commented):
Left some more minor comments, but I think this looks pretty good to me, modulo some clean ups of commented out parts, etc!

stefan-mysten commented:
@amnn The reason why in some `warn/info` we used `[Response]` or `[Query] or `[Cost]` is that all of these come from the logger, so one will get this: 1) first is the query, 2) is the validation, 3) is the response. As long as we know about it, it's fine, we can probably get the query id of an error, and then get all logs for that error, and then if we know the order we can deduct which is which. Visually, it's easier to parse and match the log to what part of the execution it happened.
<img width="1720" alt="image" src="https://github.com/MystenLabs/sui/assets/135084671/37c01c98-b9d7-4ac5-b569-3fcf3a58d67f">

amnn commented:
The tags on logger messages are fine, I just don't want that to be used everywhere for the reason I mentioned.

stefan-mysten commented:
@amnn @wlmyng thanks for the first round of reviews. I'd appreciate another review when you can. Since then, I did the following:
* addressed all the suggestions
* added `inflight_requests` metric
* added the `num_queries_top_level` that records how many times a top level node is requested in a query, after talking to @suiwombat. In a nutshell, in a query as this one:
```
{
  epoch {
    epochId
  }
  transactionBlocks {
    nodes {
      digest
    }
  }
  chainIdentifier
}
```
we'll get 
```
num_queries_top_level{path="chainIdentifier"} 1
num_queries_top_level{path="epoch"} 1
num_queries_top_level{path="transactionBlocks"} 1
```
 
* Revisited the `query_id` and `Uuid` stuff, and got rid of the abstraction. Every query gets a unique identifier and can get accessed directly from context.
* Added `query_payload_too_large_size` metric, which tracks how big the payloads are for queries that are not allowed to go through due to the too large payload.

suiwombat commented on crates/sui-graphql-rpc/src/data/pg.rs:
```diff
@@ -12,9 +14,12 @@ use diesel::{
 pub(crate) struct PgExecutor {
     pub inner: IndexerReader,
     pub limits: Limits,
+    pub metrics: Option<Metrics>,
```
is there a time when we wouldn't want to have metrics passed in? if it was required and not an option, the use/recording below would be easier.

suiwombat commented on crates/sui-graphql-rpc/src/extensions/logger.rs:
```diff
@@ -125,6 +127,7 @@ impl Extension for LoggerExtension {
         let resp = next.run(ctx, operation_name).await;
         if resp.is_err() {
             for err in &resp.errors {
+                let error_code = &err.extensions.as_ref().and_then(|x| x.get("code"));
```
Consider adding an UNKNOWN enum value and using that as a default.  Then you can use a match block much easer below instead of the "is it a code::INTERNALSERVER_ERROR else some other thing" logic.

https://github.com/MystenLabs/sui/blob/7bc4de5c00079aab15dc6b8aefaf1a6c78547156/crates/sui-graphql-rpc/src/error.rs#L14-L19

stefan-mysten replied:
Good idea, added!

suiwombat commented on crates/sui-graphql-rpc/src/extensions/logger.rs:
```diff
@@ -140,29 +143,76 @@ impl Extension for LoggerExtension {
-                        target: "async-graphql",
-                        "[Response] path={} message={}", path, err.message,
-                    );
+                    if let Some(async_graphql_value::ConstValue::String(val)) = error_code {
```
if you add a code::UNKNOWN enum value this if block/test will be easier to reason through and become shorter.

stefan-mysten commented on crates/sui-graphql-rpc/src/metrics.rs:
```diff
@@ -1,74 +1,254 @@
+
+    /// When an error occurs, GraphQL returns a vector of PathSegments,
+    /// that we can use to construct a simplified path to the actual error.
+    pub(crate) fn get_query_path_on_error(&self, query: Vec<PathSegment>) -> String {
```
This does a reduction: [first_element.last_element] in the path.

amnn replied:
That seems like a reasonable contraction, but:

- I don't think that's what's happening here -- it looks like it's computing `first.second. ... .last`.
- You will need to handle the case where there's only one path segment.
- I would recommend avoiding the `Index` components, because they are going to introduce noise in your labels.

Also, unrelated, I just realised that this doesn't use `self` so you should make this a free/top-level function.

How about this:

```rust
pub(crate) fn query_label_for_error(query: &[PathSegment]) -> String {
    let fields: Vec<_> = query.iter()
        .filter_map(|s| if let PathSegment::Field(name) { Some(name) } else { None })
        .collect();

    match &fields[..] {
        [] => "".to_string(),
        [seg] => seg.clone(),
        [fst, .., lst] => format!("{fst}..{lst}"),
    }
}
```

stefan-mysten replied:
Thanks for spotting this, fixed.

stefan-mysten commented on crates/sui-graphql-rpc/src/metrics.rs:
```diff
@@ -1,74 +1,254 @@
+    /// The query depth
+    pub query_depth: Histogram,
+    /// The size (in bytes) of the payload that is higher than the maximum
+    pub query_payload_too_large_size: Histogram,
```
We could do that, but I prefer to have it separate as they do refer to queries that were successfully accepted and those that were not.

stefan-mysten commented on crates/sui-graphql-rpc/src/context_data/db_data_provider.rs:
No, errors are logged in the logger.

stefan-mysten commented on crates/sui-graphql-rpc/src/extensions/logger.rs:
```diff
@@ -93,8 +91,10 @@ impl Extension for LoggerExtension {
             .any(|(_, operation)| operation.node.selection_set.node.items.iter().any(|selection| matches!(&selection.node, Selection::Field(field) if field.node.name.node == "__schema")));
         if !is_schema && self.config.log_request_query {
             info!(
-                target: "async-graphql",
-                "[Query] {}: {}", self.session_id().await, ctx.stringify_execute_doc(&document, variables)
+                query_id = ctx.data_unchecked::<Uuid>().to_string(),
+                "[Query] {} {}",
```
Yep. Without `to_string` it errors.
<img width="973" alt="image" src="https://github.com/MystenLabs/sui/assets/135084671/068a2bcf-6136-4ab5-8f39-2aecf16a1c71">

stefan-mysten commented on crates/sui-graphql-rpc/src/extensions/logger.rs:
```diff
@@ -54,9 +54,7 @@ struct LoggerExtension {
-        let ip_component = ip.map(|ip| format!("{}-", ip)).unwrap_or_default();
-        let uuid_component = format!("{}", Uuid::new_v4());
-        *self.session_id.lock().await = format!("{}{}", ip_component, uuid_component);
+        *self.session_id.lock().await = ip.map(|ip| format!("{}", ip)).unwrap_or_default();
```
Fixed.

stefan-mysten commented on crates/sui-graphql-rpc/src/extensions/logger.rs:
```diff
@@ -140,29 +143,76 @@ impl Extension for LoggerExtension {
+                            err.message,
+                        );
+                    }
+                } else if let Some(async_graphql_value::ConstValue::String(val)) = error_code {
```
I rewrote this, as I think it was unnecessary.

stefan-mysten commented on crates/sui-graphql-rpc/src/extensions/logger.rs:
```diff
@@ -140,29 +143,76 @@ impl Extension for LoggerExtension {
                 _ => info!(
-                    target: "async-graphql",
-                    "[Response] {}: {}", self.session_id().await, resp.data
+                    query_id = ctx.data_unchecked::<Uuid>().to_string(),
```
Yep, fixed!

stefan-mysten commented:
> The ctx still seems to be here, but not used? https://github.com/MystenLabs/sui/pull/15696#discussion_r1458123245

It's used for query id, session id, and so on.

stefan-mysten commented on crates/sui-graphql-rpc/src/server/builder.rs:
```diff
@@ -237,29 +250,54 @@ impl ServerBuilder {
+    let elapsed = instant.elapsed().as_millis() as u64;
+    if let Some(m) = metrics {
+        m.query_latency(elapsed);
+        m.inc_num_queries();
```
Yes, absolutely.

stefan-mysten commented on crates/sui-graphql-rpc/src/metrics.rs:
```diff
@@ -1,74 +1,254 @@
+    /// The time it takes for the GraphQL service to execute the request by path
+    pub _query_latency_by_path: HistogramVec,
+    /// Number of errors by path and type.
+    /// - max_query_nodes is the number of errors due to a number of input nodes higher than max
+    /// - max_query_depth is the number of errors due to a query depth higher than max allowed
+    /// - max_output_nodes is the number of errors due to a higher number of output nodes than max
```
Ok, bundled it together.

stefan-mysten commented on crates/sui-graphql-rpc/src/metrics.rs:
Done.

stefan-mysten commented on crates/sui-graphql-rpc/src/server/builder.rs:
```diff
@@ -237,29 +250,54 @@ impl ServerBuilder {
+        }
+        m.request_metrics.inflight_requests.dec();
+
+        // TODO Is this the right way to get the top level query path names?
```
That sounds better. I wonder if there is an overhead for using this `resolve` hook instead of doing it once everything got parsed and returned. I added the hook and we can do some experiments later.

stefan-mysten replied:
```graphql
{
  epoch1: epoch {
    epochId
  }
  epoch2: epoch {
    epochId 
  }
  chainIdentifier
}
```
results in 
```
# HELP num_queries_top_level Number of queries for each top level node
# TYPE num_queries_top_level counter
num_queries_top_level{path="chainIdentifier"} 1
num_queries_top_level{path="epoch"} 2
```

stefan-mysten commented on crates/sui-graphql-rpc/src/error.rs:
```diff
@@ -14,8 +14,9 @@ use crate::context_data::db_data_provider::DbValidationError;
     pub const BAD_REQUEST: &str = "BAD_REQUEST";
     pub const BAD_USER_INPUT: &str = "BAD_USER_INPUT";
-    pub const GRAPHQL_VALIDATION_FAILED: &str = "GRAPHQL_VALIDATION_FAILED";
+    pub const _GRAPHQL_VALIDATION_FAILED: &str = "GRAPHQL_VALIDATION_FAILED";
```
Changed the error code when logging `query.len > max allowed` to `BAD_USER_INPUT`, thus this is not needed, but I kept it for now.

stefan-mysten commented on crates/sui-graphql-rpc/src/extensions/logger.rs:
```diff
@@ -41,29 +40,15 @@ pub struct Logger {
     config: LoggerConfig,
 }
 
-impl LoggerExtension {
```
We don't need these anymore, as we're using directly the context.
`ctx.data_unchecked::<SocketAddr>()`

stefan-mysten commented on crates/sui-graphql-rpc/src/extensions/logger.rs:
```diff
@@ -73,11 +58,25 @@ impl Extension for LoggerExtension {
         next.run(ctx, request).await
     }
 
+    async fn resolve(
```
Added this hook to get the path node in order to measure the top level queries that are being called.

stefan-mysten commented on crates/sui-graphql-rpc/src/extensions/query_limits_checker.rs:
```diff
@@ -222,6 +231,7 @@ impl QueryLimitsChecker {
         sel_set: &Positioned<SelectionSet>,
         cost: &mut ComponentCost,
         variables: &Variables,
+        ctx: &ExtensionContext<'_>,
```
We need this to get the `query id` and the `session_id`.

amnn commented on crates/sui-graphql-rpc/src/extensions/logger.rs:
```diff
@@ -123,8 +116,11 @@ impl Extension for LoggerExtension {
+        let query_id = ctx.data_unchecked::<Uuid>().to_string();
         if resp.is_err() {
             for err in &resp.errors {
+                let error_code = &err.extensions.as_ref().and_then(|x| x.get("code"));
```
I was initially going to suggest a further refinement of this, but I think it's probably not worth it, because the type at this point is not `Option<&str>` it's `Option<Value>`, so it's going to be a pain to create the version of this that contains `UNKNOWN`.

amnn commented on crates/sui-graphql-rpc/src/extensions/logger.rs:
```diff
@@ -141,29 +137,43 @@ impl Extension for LoggerExtension {
-                        target: "async-graphql",
-                        "[Response] message={}", err.message,
-                    );
+                    if let Some(async_graphql_value::ConstValue::String(val)) = error_code {
```
nit: Call this `error_code`, and then you can also turn `error_code = val` into just `error_code` below?

amnn commented on crates/sui-graphql-rpc/src/extensions/logger.rs:
```diff
@@ -141,29 +137,43 @@ impl Extension for LoggerExtension {
+                            query_id,
+                            session_id,
+                            error_code = code::UNKNOWN,
+                            "[Response] message={}",
```
Do we need to add a `path` here as well?

stefan-mysten replied:
Good catch!

amnn commented on crates/sui-graphql-rpc/src/extensions/logger.rs:
```diff
@@ -141,29 +137,43 @@ impl Extension for LoggerExtension {
-                        target: "async-graphql",
-                        "[Schema] {}: {}", self.session_id().await, resp.data
-                    );
+                    debug!(query_id, session_id, "[Response] {}", resp.data);
```
nit: This one was tagged with `Schema` before -- was the change to `[Response]` before intentional?

stefan-mysten replied:
Oops, my mistake. Fixed!

amnn commented on crates/sui-graphql-rpc/src/extensions/logger.rs:
```diff
@@ -123,8 +116,11 @@ impl Extension for LoggerExtension {
         next: NextExecute<'_>,
     ) -> Response {
         let resp = next.run(ctx, operation_name).await;
+        let session_id = ctx.data_unchecked::<SocketAddr>().to_string();
+        let query_id = ctx.data_unchecked::<Uuid>().to_string();
```
Ascribe the type on the variable, rather than using the "rocket fish" operator.  Also, I think you don't need to convert this to a string -- `tracing` will do this for you.
```suggestion
        let session_id: SocketAddr = ctx.data_unchecked();
        let query_id: Uuid = ctx.data_unchecked();
```

stefan-mysten replied:
It will not work without `to_string()`.

amnn replied:
The piece that was missing here, was that you need to use a `%` sigil to get it to `Display` the type, details here:

https://docs.rs/tracing/latest/tracing/#recording-fields

stefan-mysten replied:
Thanks, that's quite inconspicuous! I will put up another PR to fix this.

amnn commented on crates/sui-graphql-rpc/src/extensions/query_limits_checker.rs:
```diff
@@ -135,13 +120,28 @@ impl Extension for QueryLimitsChecker {
         variables: &Variables,
         next: NextParseQuery<'_>,
     ) -> ServerResult<ExecutableDocument> {
+        let query_id = ctx.data_unchecked::<Uuid>().to_string();
+        let session_id = ctx.data_unchecked::<SocketAddr>().to_string();
+        let metrics = ctx.data_unchecked::<Metrics>();
```
Same comment here on ascribing the type on the variable and avoid the `to_string()`.

```suggestion
        let query_id: Uuid = ctx.data_unchecked();
        let session_id: SocketAddr = ctx.data_unchecked();
        let metrics: Metrics = ctx.data_unchecked();
```

stefan-mysten replied:
Same as before, it will not work for the `query_id` and `session_id` variables.

amnn commented on crates/sui-graphql-rpc/src/extensions/query_limits_checker.rs:
```diff
@@ -326,34 +336,60 @@ impl QueryLimitsChecker {
+    pos: Option<Pos>,
+    ctx: &ExtensionContext<'_>,
+) -> ServerResult<()> {
+    let query_id = ctx.data_unchecked::<Uuid>().to_string();
+    let session_id = ctx.data_unchecked::<SocketAddr>().to_string();
```
(ditto)

stefan-mysten replied:
Same as before, it will not work for the query_id and session_id variables as tracing does not seem to be able to convert Uuid into string.

amnn commented on crates/sui-graphql-rpc/src/extensions/query_limits_checker.rs:
```diff
@@ -326,34 +336,60 @@ impl QueryLimitsChecker {
+    let error_code = code::BAD_USER_INPUT;
     if cost.input_nodes > limits.max_query_nodes {
-        return Err(ServerError::new(
+        info!(
```
optional, but these things seem vaguely macro-able (to me)

amnn commented on crates/sui-graphql-rpc/src/server/builder.rs:
```diff
@@ -640,39 +698,106 @@ pub mod tests {
-        assert_eq!(metrics2.input_nodes.get_sample_sum(), 1.);
-        assert_eq!(metrics2.output_nodes.get_sample_sum(), 1.);
-        assert_eq!(metrics2.query_depth.get_sample_sum(), 1.);
+        // TODO these do not get properly triggered. Need more investigation, seems to be
```
Flagging this part for clean-up -- but I'm not sure what the right thing to do is here -- will need someone who is familiar with `mysten-metrics` to have a look.

stefan-mysten replied:
I scheduled some time with @suiwombat to look together at this, and will ping other folks as needed.

amnn reviewed (approved):
This looks good, thanks for persisting with it @stefan-mysten !  Accepting with comments -- mostly nits, but two big things to look at:

- Cleaning up the tests, and getting someone who's familiar with `mysten-metrics` to help you out there.
- The error label function -- what you've described it as in the PR comment seems reasonable, but the implementation doesn't seem to match it.

stefan-mysten commented on crates/sui-graphql-rpc/src/extensions/logger.rs:
```diff
@@ -93,8 +82,10 @@ impl Extension for LoggerExtension {
         if !is_schema && self.config.log_request_query {
             info!(
-                target: "async-graphql",
-                "[Query] {}: {}", self.session_id().await, ctx.stringify_execute_doc(&document, variables)
+                query_id = ctx.data_unchecked::<Uuid>().to_string(),
+                session_id = ctx.data_unchecked::<SocketAddr>().to_string(),
+                "[Query] {}",
+                ctx.stringify_execute_doc(&document, variables)
             );
```
I did, but I think because I resolved the previous comments you might have not seen it. It does not work because the trait `tracing::Value` is not implemented for `uuid::Uuid`.  
https://github.com/MystenLabs/sui/pull/15696#discussion_r1458613775
